File: python_code/playback_scraper.py
import os
import time
import json
import glob
from dotenv import load_dotenv
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager
import auth

# Load environment variables
load_dotenv()

def sanitize_filter_name(filter_name):
    """Sanitize filter name for use in filenames and directories"""
    import unicodedata
    normalized = unicodedata.normalize('NFD', filter_name)
    without_accents = ''.join(char for char in normalized if unicodedata.category(char) != 'Mn')
    sanitized = ''.join(c if c.isalnum() else '_' for c in without_accents).lower()
    return sanitized

def setup_driver():
    """Set up Chrome WebDriver"""
    options = Options()
    options.add_argument("--headless")  # Run headless for playback scraping
    options.add_argument("--disable-gpu")
    options.add_argument("--no-sandbox")
    
    service = Service(ChromeDriverManager().install())
    driver = webdriver.Chrome(service=service, options=options)
    return driver

def scrape_playback(driver, playback_url):
    """Scrape video URLs from playback page"""
    print(f"   Scraping playback: {playback_url[:60]}...")
    
    try:
        driver.get(playback_url)
        time.sleep(5)  # Wait for page to load
        
        # Find video elements
        videos = []
        
        # Try finding video elements by source tags
        video_elements = driver.find_elements(By.TAG_NAME, "video")
        for video in video_elements:
            sources = video.find_elements(By.TAG_NAME, "source")
            for source in sources:
                src = source.get_attribute("src")
                if src and src.endswith('.webm'):
                    videos.append(src)
        
        # Also check for direct links to videos
        links = driver.find_elements(By.TAG_NAME, "a")
        for link in links:
            href = link.get_attribute("href")
            if href and ('.webm' in href or 'video' in href):
                if href not in videos:
                    videos.append(href)
        
        return list(set(videos))  # Remove duplicates
        
    except Exception as e:
        print(f"   ✗ Error scraping playback: {e}")
        return []

def load_recordings(bbb_filter=''):
    """Load recordings from session data"""
    # Determine search directory
    if bbb_filter:
        safe_name = sanitize_filter_name(bbb_filter)
        search_dir = f"scraped_data/{safe_name}"
    else:
        search_dir = "scraped_data"
    
    # Find session files
    session_files = glob.glob(f"{search_dir}/session_*.json")
    if not session_files:
        print(f"No session files found in {search_dir}")
        return []
    
    # Load the most recent session file
    latest_file = max(session_files, key=os.path.getctime)
    print(f"Loading: {latest_file}")
    
    with open(latest_file, 'r', encoding='utf-8') as f:
        return json.load(f)

def load_existing_playback_data(search_dir):
    """Load existing playback data to avoid re-scraping"""
    try:
        if not os.path.exists(search_dir):
            return {}
        
        # Find all playback files
        playback_files = glob.glob(f"{search_dir}/playback_data_*.json")
        if not playback_files:
            return {}
        
        # Sort by modification time (newest first)
        playback_files.sort(key=os.path.getctime, reverse=True)
        
        latest_file = playback_files[0]
        print(f"Loading existing data from: {latest_file}")
        
        with open(latest_file, 'r', encoding='utf-8') as f:
            data = json.load(f)
            
        # Map playback_link -> data
        existing_map = {}
        for item in data:
            if 'playback_link' in item and 'scraped_content' in item:
                existing_map[item['playback_link']] = item
        
        print(f"✓ Loaded {len(existing_map)} existing recordings")
        return existing_map
        
    except Exception as e:
        print(f"⚠ Could not load existing data: {e}")
        return {}

def main():
    """Main function"""
    print("Starting SENCE Playback Scraper (Python)...\n")
    
    # Get filter from environment
    bbb_filter = os.getenv('BBB_FILTER', '')
    
    # Determine directories
    if bbb_filter:
        safe_name = sanitize_filter_name(bbb_filter)
        output_dir = f"scraped_data/{safe_name}"
    else:
        output_dir = "scraped_data"
        
    # Load recordings to process
    recordings = load_recordings(bbb_filter)
    if not recordings:
        print("No recordings to process.")
        return
    
    # Load existing data
    existing_map = load_existing_playback_data(output_dir)
    
    # Identify new recordings
    new_recordings = [r for r in recordings if r['playback_link'] not in existing_map]
    print(f"\nStatus: {len(existing_map)} existing, {len(new_recordings)} new\n")
    
    # Setup driver only if needed
    if new_recordings:
        driver = setup_driver()
        print("Browser started for scraping new recordings...")
    else:
        driver = None
        print("No new recordings to scrape.")
    
    try:
        enriched_data = []
        
        for i, recording in enumerate(recordings, 1):
            print(f"[{i}/{len(recordings)}] {recording.get('name', 'Unknown')[:60]}...")
            playback_link = recording.get('playback_link', '')
            
            # Check existing
            if playback_link in existing_map:
                print("   ✓ Using cached data")
                enriched_data.append(existing_map[playback_link])
                continue
            
            # Scrape new
            if not playback_link:
                print("   ⚠ No playback link")
                continue
                
            videos = scrape_playback(driver, playback_link)
            
            if videos:
                print(f"   ✓ Found {len(videos)} video(s) (Scraped)")
                enriched_data.append({
                    "name": recording['name'],
                    "playback_link": playback_link,
                    "scraped_content": {
                        "videos": videos
                    }
                })
            else:
                print("   ⚠ No videos found")

        # Save results
        os.makedirs(output_dir, exist_ok=True)
        
        timestamp = time.strftime("%Y-%m-%dT%H-%M-%S")
        filename = f"{output_dir}/playback_data_{timestamp}.json"
        
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(enriched_data, f, indent=2, ensure_ascii=False)
        
        print(f"\n✓ Saved {len(enriched_data)} results to {filename}")
        
    except Exception as e:
        print(f"\n✗ An error occurred: {e}")
        import traceback
        traceback.print_exc()
    finally:
        if driver:
            print("\nClosing browser...")
            driver.quit()

if __name__ == "__main__":
    main()

File: playback_scraper.js
import 'dotenv/config';
import puppeteer from 'puppeteer';
import fs from 'fs-extra';
import path from 'path';
import { autoLogin } from './auth.js';

const CONFIG = {
    headless: false,
    defaultViewport: null,
    args: ['--start-maximized']
};

// Helper: Load all session files and deduplicate recordings
async function loadRecordings() {
    // Determine search directory based on BBB_FILTER
    const filter = process.env.BBB_FILTER;
    let searchDir = '.';
    
    if (filter) {
        const safeName = filter
            .normalize("NFD").replace(/[\u0300-\u036f]/g, "")
            .replace(/[^a-z0-9]/gi, '_')
            .toLowerCase(); // Match session_scraper.js sanitization
        searchDir = `scraped_data/${safeName}`;
        
        if (!await fs.pathExists(searchDir)) {
            console.log(`No session data folder found at ${searchDir}`);
            return [];
        }
    }
    
    const files = await fs.readdir(searchDir);
    const sessionFiles = files.filter(f => f.startsWith('session_') && f.endsWith('.json') && f !== 'session_cookies.json');
    
    if (sessionFiles.length === 0) {
        console.log("No session_data*.json files found");
        return [];
    }

    console.log(`Found ${sessionFiles.length} session files`);
    
    const allRecordings = [];
    for (const file of sessionFiles) {
        try {
            const filePath = path.join(searchDir, file);
            const { bbb_session: { recordings = [] } } = await fs.readJson(filePath);
            console.log(` - ${file}: ${recordings.length} recordings`);
            allRecordings.push(...recordings);
        } catch (e) {
            console.error(`  ✗ Error reading ${file}:`, e.message);
        }
    }
    
    // Deduplicate by playback_link
    const unique = allRecordings.filter((r, i, arr) => 
        arr.findIndex(x => x.playback_link === r.playback_link) === i
    );
    
    console.log(`✓ Total unique recordings: ${unique.length}\n`);
    return unique;
}

// Helper: Extract media sources from page
async function extractMediaSources(page) {
    return await page.evaluate(() => {
        const videos = new Set();
        const audios = new Set();
        
        // Collect from video tags
        document.querySelectorAll('video').forEach(v => {
            if (v.src) videos.add(v.src);
            v.querySelectorAll('source').forEach(s => s.src && videos.add(s.src));
        });
        
        // Collect from audio tags
        document.querySelectorAll('audio').forEach(a => {
            if (a.src) audios.add(a.src);
            a.querySelectorAll('source').forEach(s => s.src && audios.add(s.src));
        });
        
        // Collect orphan sources
        document.querySelectorAll('source').forEach(s => {
            if (!s.src) return;
            const parent = s.parentElement?.tagName;
            if (parent === 'VIDEO') videos.add(s.src);
            else if (parent === 'AUDIO') audios.add(s.src);
            else videos.add(s.src); // Fallback
        });
        
        return {
            videos: [...videos],
            audios: [...audios],
            slide_count: document.querySelectorAll('svg').length
        };
    });
}

// Helper: Load existing playback data to avoid re-scraping
async function loadExistingPlaybackData(searchDir) {
    try {
        if (!await fs.pathExists(searchDir)) return new Map();
        
        const files = await fs.readdir(searchDir);
        const playbackFiles = files.filter(f => f.startsWith('playback_data_') && f.endsWith('.json'));
        
        if (playbackFiles.length === 0) return new Map();
        
        // Sort by time (newest first) to get latest data
        playbackFiles.sort((a, b) => {
            const timeA = fs.statSync(path.join(searchDir, a)).mtime.getTime();
            const timeB = fs.statSync(path.join(searchDir, b)).mtime.getTime();
            return timeB - timeA;
        });
        
        // Load the latest file
        const latestFile = path.join(searchDir, playbackFiles[0]);
        console.log(`Loading existing data from ${playbackFiles[0]}`);
        const data = await fs.readJson(latestFile);
        
        // Map playback_link -> scraped data
        const map = new Map();
        data.forEach(item => {
            if (item.playback_link && item.scraped_content) {
                map.set(item.playback_link, item);
            }
        });
        
        console.log(`✓ Loaded ${map.size} existing recordings`);
        return map;
    } catch (e) {
        console.log(`⚠ Could not load existing data: ${e.message}`);
        return new Map();
    }
}

async function main() {
    console.log("Starting SENCE Playback Scraper...\n");

    try {
        const recordings = await loadRecordings();
        
        if (recordings.length === 0) {
            console.log("No recordings to process");
            return;
        }

        // Determine output directory to check for existing data
        const filter = process.env.BBB_FILTER;
        let outputDir = 'scraped_data';
        if (filter) {
            const safeName = filter
                .normalize("NFD").replace(/[\u0300-\u036f]/g, "")
                .replace(/[^a-z0-9]/gi, '_')
                .toLowerCase();
            outputDir = `scraped_data/${safeName}`;
        }

        // Load existing data
        const existingDataMap = await loadExistingPlaybackData(outputDir);
        
        // Identify new recordings
        const newRecordings = recordings.filter(rec => !existingDataMap.has(rec.playback_link));
        console.log(`\nStatus: ${existingDataMap.size} existing, ${newRecordings.length} new\n`);

        const browser = await puppeteer.launch(CONFIG);
        const page = await browser.newPage();
        const enrichedData = [];

        try {
            // Only authenticate if we have new recordings to scrape
            if (newRecordings.length > 0) {
                console.log("Authenticating...");
                // Use the first new recording link for auth, or fallback to first recording
                const authLink = newRecordings[0].playback_link;
                await page.goto(authLink, { waitUntil: 'networkidle2' });
                
                if (!await autoLogin(page)) {
                    console.log("⚠ Auto-login incomplete - session may be required\n");
                }
            }

            // Process all recordings (use existing or scrape new)
            for (let i = 0; i < recordings.length; i++) {
                const rec = recordings[i];
                console.log(`[${i + 1}/${recordings.length}] ${rec.name.substring(0, 40)}...`);
                
                // Check if we already have data
                if (existingDataMap.has(rec.playback_link)) {
                    console.log("  ✓ Using cached data");
                    enrichedData.push(existingDataMap.get(rec.playback_link));
                    continue;
                }

                // Scrape new recording
                try {
                    await page.goto(rec.playback_link, { waitUntil: 'networkidle2', timeout: 60000 });
                    
                    // Wait for media elements
                    try {
                        await page.waitForSelector('video, audio, source', { timeout: 10000 });
                    } catch (e) {
                        console.log("  ⚠ No media tags found");
                    }
                    
                    await new Promise(r => setTimeout(r, 2000)); // Stability wait
                    
                    const scrapedContent = await extractMediaSources(page);
                    
                    if (scrapedContent.videos.length > 0) {
                        console.log(`  ✓ ${scrapedContent.videos.length} video(s) (Scraped)`);
                        enrichedData.push({
                            ...rec,
                            realplayback_url: page.url(),
                            scraped_content: scrapedContent
                        });
                    } else {
                        console.log("  ⚠ No videos found");
                    }
                    
                } catch (e) {
                    console.error(`  ✗ Error: ${e.message}`);
                }
            }
            
            // Save results
            const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
            let outFile = `playback_data_${timestamp}.json`;
            
            // Create output directory if BBB_FILTER is set
            if (filter) {
                const safeName = filter
                    .normalize("NFD").replace(/[\u0300-\u036f]/g, "")
                    .replace(/[^a-z0-9]/gi, '_')
                    .toLowerCase();
                const outputDir = `scraped_data/${safeName}`;
                await fs.ensureDir(outputDir);
                outFile = `${outputDir}/${outFile}`;
            }
            
            await fs.writeJson(outFile, enrichedData, { spaces: 4 });
            console.log(`\n✓ Saved to ${outFile}`);
            
        } finally {
            await browser.close();
        }

    } catch (error) {
        console.error("✗ Error:", error.message);
        if (process.argv.includes('--debug')) {
            console.log("Debug screenshot saved");
        }
    }
}

main();
